refactor: model/tests: Simplify Zulip Feature Level Handling.

rsashank opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
Fixes #1364
Pure integer representation replacing the current None value when ZFL is absent.


### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- [ ] 

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [ ] Discussed in **#zulip-terminal** in `topic`
- [x] Fully fixes #1364
- [ ] Partially fixes issue #
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [ ] Manually - Behavioral changes
- [ ] Manually - Visual changes
- [x] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [x] It individually passes linting and tests
- [x] It contains test additions for any new behavior
- [x] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->

neiljp reviewed (commented):
@rsashank Thanks for the moving the previous PR forward :+1:

This certainly addresses some of the earlier points in that PR, though I did note that you dropped one change when comparing to it.

This turned out to be more work than I anticipated to complete fully, since we use various different names around the repo, including at least:
- zulip_feature_level (server field name)
- ZFL (abbreviation)
- server_feature_level (in model, to distinguish from client)
- feature_level (short variable, in tests)

That variation in names, and that they're not all code, certainly hasn't helped. Grepping through the repo, it seems there are a few outstanding unchanged parts remaining.

When you rebase and push this commit again, I suspect this will fail gitlint locally or in CI. A few other notes in that direction:
- Fixes lines are good near the bottom after a blank line
- This updates more than just the model, but you don't need to mention tests in the commit title (it's assumed they're always updated where necessary)
- Since this builds upon the other PR it would be good to add a co-authored-by field near the bottom of the commit (see `git log|grep -i co-au` for examples)
- For commit text, describing what the commit changes can be good, but just as important (if not more) is the reasoning (why is integer good? why is removing None good? what did None represent? when is ZFL absent?)

rsashank commented on tests/model/test_model.py:
```diff
@@ -3427,7 +3427,7 @@ def test__handle_subscription_event_visual_notifications(
             (
                 {"op": "peer_add", "stream_id": 99, "user_id": 12},
-                None,
+                0,
```
I haven't directly created a branch off the previous PR; instead, I've looked at those changes and feedback to make changes in a new branch. Oversight on my part, will fix this and submit a PR soon.

rsashank commented on zulipterminal/model.py:
```diff
@@ -586,7 +586,7 @@ def update_stream_message(
             request["content"] = content
 
-        if self.server_feature_level is not None and self.server_feature_level >= 9:
+        if 9 <= self.server_feature_level != 0:
```
Oops.

rsashank commented on tests/model/test_model.py:
```diff
@@ -278,7 +278,7 @@ def test_register_initial_desired_events(self, mocker, initial_data):
                 10,
                 {1: "Indefinite [Organization default]"},
-                id="ZFL=None_no_stream_retention_realm_retention=None",
+                id="ZFL=10_no_stream_retention_realm_retention=None",
```
Added the ZFL=0 case with the same parameters as ZFL=10.

rsashank commented on tests/model/test_model.py:
```diff
@@ -3255,11 +3262,11 @@ def test__handle_typing_event(
             ),
         ],
         ids=[
-            "remove_18_in_home_view:already_unmuted:ZFLNone",
-            "remove_19_in_home_view:muted:ZFLNone",
-            "add_19_in_home_view:already_muted:ZFLNone",
-            "add_30_in_home_view:unmuted:ZFLNone",
-            "remove_30_is_muted:already_unmutedZFL139",
+            "remove_18_in_home_view:already_unmuted:ZFL0",
+            "remove_19_in_home_view:muted:ZFL0",
+            "add_19_in_home_view:already_muted:ZFL0",
+            "add_30_in_home_view:unmuted:ZFL0",
+            "remove_30_is_muted:already_unmuted:ZFL139",
```
The absence of a ":" after already_unmuted seemed inconsistent with other ids. 
While this change is unrelated to the commit/PR, it appears too minor to be a separate commit. Is that fine?

rsashank replied:
The absence of a ":" after already_unmuted seemed inconsistent with other ids. (line 3269)
While this change is unrelated to the commit/PR, it appears too minor to be a separate commit. Is that fine?

rsashank commented on tests/model/test_model.py:
```diff
@@ -3255,11 +3262,11 @@ def test__handle_typing_event(
             ),
         ],
         ids=[
-            "remove_18_in_home_view:already_unmuted:ZFLNone",
-            "remove_19_in_home_view:muted:ZFLNone",
-            "add_19_in_home_view:already_muted:ZFLNone",
-            "add_30_in_home_view:unmuted:ZFLNone",
-            "remove_30_is_muted:already_unmutedZFL139",
+            "remove_18_in_home_view:already_unmuted:ZFL0",
+            "remove_19_in_home_view:muted:ZFL0",
+            "add_19_in_home_view:already_muted:ZFL0",
+            "add_30_in_home_view:unmuted:ZFL0",
```
I found these ones that weren't changed initially in the PR.

I'm still looking to identify any additional code that may need adjustments.

rsashank commented:
Thanks for the detailed feedback! @neiljp 

I've updated the PR, and I believe all of them have been addressed. I couldn't find any outliers, but please correct me if I'm mistaken. (Checked using VSCode, searched for ZFL and feature_level). I'll further check with grep as well.

Rebased and updated the commit message as per feedback as well.

@zulipbot add "PR needs review"
@zulipbot remove "PR awaiting update"

mounilKshah reviewed (commented):
Thank you @rsashank for updating the PR
I noticed that you have implemented all the suggestions given by @neiljp 

As Neil mentioned in the PR feedback, we have addressed ZFL with multiple phrases like feature level, Zulip feature level, ZFL etc. in code comments or in test case IDs. How about we also add another commit to this PR which renames them to the same name? What do you think of this idea @neiljp and @rsashank ?

rsashank commented:
@mounilKshah That makes sense but what's the best way to ensure uniformity in variable names?

Would simply adding `_ZFL` to the end of variables where needed be sufficient (so you can grep through the repo with ZFL), or should we consider a more uniform rewrite across the repo?

mounilKshah commented:
I was referring to comments and test IDs to have a uniform naming pattern across the repository, variables need not be changed. Naming all instances as _ZFL_ can confuse a new contributor about its meaning, so maybe we can use the entire phrase _Zulip Feature Level_.

rsashank commented:
Oh my bad. Makes sense, @neiljp should I make another commit addressing this?

neiljp reviewed (commented):
@rsashank Thanks for the update :+1:

This looks to have full coverage for the change, so I've just pushed back with a slightly adjusted commit message.

I'd like to merge this soon (once CI passes), since tests written in other PRs assume `None` is possible, so setting this new expectation would be useful!

For that reason I'd prefer to consider any followup on this refactor separately.
- While abbreviations aren't great in general, expanding ZFL would make for very long test case ids or comments (and it is very common)
- I think we could certainly expand some versions to a common form, eg. feature_level
- Documenting what ZFL corresponds to would likely be well-placed in api_types.py? We shouldn't use it in docs without the expanded form first, or point to api_types perhaps.
- We could rename ZFL/zulip_feature_level to reflect the server-provided name, ie. server_feature_level, but that would be less important in terms of making it easier to search for phrases; it may also be something we adjust if we namespace server and organization names, which has been discussed as a possible migration in the server already.
